homeassistant/components/somfy/config_flow.py

```diff
@@ -0,0 +1,187 @@
+    }
+
+
+@config_entries.HANDLERS.register('somfy')
+class SomfyFlowHandler(config_entries.ConfigFlow):
+    """Handle a config flow."""
+
+    VERSION = 1
+    CONNETION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    def __init__(self):
+        """Initialize flow."""
+        self.flow_impl = None
+
+    async def async_step_import(self, user_input=None):
+        """Handle external yaml configuration."""
+        if self.hass.config_entries.async_entries(DOMAIN):
+            return self.async_abort(reason='already_setup')
+
+        self.flow_impl = DOMAIN
+
+        return await self.async_step_auth()
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow start."""
+        flows = self.hass.data.get(DATA_FLOW_IMPL, {})
+
+        if self.hass.config_entries.async_entries(DOMAIN):
+            return self.async_abort(reason='already_setup')
+
+        if not flows:
+            _LOGGER.debug("no flows")
+            return self.async_abort(reason='no_flows')
+
+        if len(flows) == 1:
+            self.flow_impl = list(flows)[0]
+            return await self.async_step_auth()
+
+        if user_input is not None:
+            self.flow_impl = user_input['flow_impl']
+            return await self.async_step_auth()
+
+        return self.async_show_form(
+            step_id='user',
+            data_schema=vol.Schema({
+                vol.Required('flow_impl'):
+                    vol.In(list(flows))
+            }))
+
+    async def async_step_auth(self, user_input=None):
+        """Create an entry for auth."""
+        if self.hass.config_entries.async_entries(DOMAIN):
+            return self.async_abort(reason='external_setup')
+
+        errors = {}
+
+        if user_input is not None:
+            errors['base'] = 'follow_link'
+
+        try:
+            with async_timeout.timeout(10):
+                url = await self._get_authorization_url()
+        except asyncio.TimeoutError:
+            return self.async_abort(reason='authorize_url_timeout')
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected error generating auth url")
+            return self.async_abort(reason='authorize_url_fail')
+
+        return self.async_show_form(
+            step_id='auth',
+            description_placeholders={'authorization_url': url},
+            errors=errors,
+        )
+
+    async def _get_authorization_url(self):
+        """Get Somfy authorization url."""
+        from pymfy.api.somfy_api import SomfyApi
+        flow = self.hass.data[DATA_FLOW_IMPL][self.flow_impl]
+        client_id = flow[CLIENT_ID]
+        client_secret = flow[CLIENT_SECRET]
+        redirect_uri = '{}{}'.format(
+            self.hass.config.api.base_url, AUTH_CALLBACK_PATH)
+        api = SomfyApi(client_id, client_secret, redirect_uri)
+
+        self.hass.http.register_view(SomfyAuthCallbackView())
+        return api.get_authorization_url()
```

Comment:
Is this async safe?

Reply:
Same here, no I/O.

Reply:
For, sure there is no network operation here.

Reply:
This could block since it uses `SystemRandom` which uses `os.urandom` in the oauth library.

https://docs.python.org/3/library/os.html#os.urandom

Reply:
I think that it's fine to use urandom in an async context, worst case it will block a bit of startup. Always going out of async is slow.